### ISSUE TITLE
console defaults to using nodePort when opening links to jenkins et all

### ISSUE THREAD

**Issue by rawlingsj** [fabric8io/fabric8#6290]
Even though we're using ingress the fabric8 console is still trying to use the service ip and nodeport in the project dashboard links
![screenshot 2016-08-26 13 10 58](https://cloud.githubusercontent.com/assets/1457180/18004867/90b0a852-6b8e-11e6-83e1-6f310a89b386.png)
and tool bar links

![screenshot 2016-08-26 13 10 10](https://cloud.githubusercontent.com/assets/1457180/18004849/72d5e522-6b8e-11e6-9105-089f4f92bdb2.png)

**Comment by gashcrumb** [comment]
Yep, so currently there's no code in place to use ingress instead of the kubernetes service node/port, this will take a bit of work to get implemented.

**Comment by gashcrumb** [comment]
I think maybe we can just check if there's a matching ingress when we're building/updating the model an use that to make the links rather than the service node and port.

**Comment by gashcrumb** [comment]
@rawlingsj Having a look at the current ingress examples it seems like it's going to be pretty difficult and error prone to definitively come up with a canonical URL client-side.  I wonder if we're creating ingresses for our services, would it make more sense to use an annotation that the console would use for these links instead?

**Comment by rawlingsj** [comment]
@gashcrumb ok sure, that can be done, so add an annotation to each kubernetes service of the external / ingress hostname?

**Comment by jstrachan** [comment]
Actually one thing thats really bugged me for the longest time is all this crazy complex code we've had in the console or java libraries (or kubernetes pipeline stuff) to figure out external URLs for things. 

It'd be much much simpler, now we've got the exposecontroller, for it to just annotate what it thinks is the right answer for how folks should access services if its watching services anyways!

Then we can get all our code to look for some magic annotation or two and use that above all other things; then folks can always manually set magic URLs if they want; or exposecontroller will do its best too etc

**Comment by jstrachan** [comment]
e.g. the console shouldn't have to watch for a router (in any namespace anywhere - maybe where the current user can see) and Routes; or ingress controller (again could be in any namespace) plus Ingress - lets let the exposecontroller do that logic and just annotate the services with what it thinks is the right external URL (nodePort, LoadBalancer IP, ingress, route etc)

**Comment by rawlingsj** [comment]
Sounds great!  Should be straight forward - got a preference for the annotation name?

**Comment by rawlingsj** [comment]
If you want I can add that in now

**Comment by jstrachan** [comment]
lets go for it! fabric8.io/exposeUrl or something? don't mind really! :)

**Comment by rawlingsj** [comment]
sounds good, I'll add it in and release it

**Comment by rawlingsj** [comment]
That's part done, I couldn't find out where to get the host IP address when adding the nodePort URLs.  It might be something we add to the exposecontroller config map during deploy if its not available on the node resource.  I can do that in the morning but for now an example is..

```
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      fabric8.io/exposeUrl: :30639
```

or when ingress is used

```
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      fabric8.io/exposeUrl: fabric8.default.test.fabric8.io
```

I've just thought, I didn't add the protocol, do we tend to use http/https in the name of a port?  If we do I can check that and default to http if no name found?

**Comment by gashcrumb** [comment]
Yeah, ideally we'd want the protocol if known, maybe via a separate
annotation or something, but we can default to http for starters I guess.
Less muckery we have to do client side the better though :-)

**Comment by jstrachan** [comment]
@rawlingsj to get the host IP you iterate through the Nodes like this Java code: https://github.com/fabric8io/fabric8/blob/master/components/kubernetes-api/src/main/java/io/fabric8/kubernetes/api/KubernetesHelper.java#L1434. 

Its maybe worth a code review of the logic we've been using in Java and hawtio to check we've all the bases covered:

* Java code: https://github.com/fabric8io/fabric8/blob/master/components/kubernetes-api/src/main/java/io/fabric8/kubernetes/api/KubernetesHelper.java#L1296
* hawtio code: https://github.com/hawtio/hawtio-kubernetes/blob/e5327b53ebae4e473fc0f1b9b2dd9370ebe081e6/plugins/kubernetes/ts/kubernetesHelpers.ts#L591

Its surprisingly complex really ;). Whats really nice about the exposecontroller approach is we've one place to put all this complex logic; plus we can use configuration (a ConfigMap) to customize how the discovery works; plus folks could always edit the annotation directly etc.

BTW we should maybe add an annotation to denote cases where folks have provided their own custom exposeUrl? e.g. `fabric8.io/exposeUrlConfigured: true` or something to tell `exposecontroller` not to re-calculate the URL as the user has customized it?

**Comment by rawlingsj** [comment]
Ok so gofabric8 0.4.47 https://github.com/fabric8io/gofabric8/releases/tag/v0.4.47
and pull the latest fabric8/exposecontroller:latest image each kubernetes service will have the external URL as an annotation.  e.g.

if node ports are used
```
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      fabric8.io/exposeUrl: http://192.168.99.100:30639
```
or openshift routes
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    fabric8.io/exposeUrl: http://fabric8.test.fabric8.io
```

**Comment by gashcrumb** [comment]
@rawlingsj awesome, will give it a go...

**Comment by jstrachan** [comment]
@gashcrumb @rawlingsj is this fixed now?

**Comment by gashcrumb** [comment]
I think it is, hoping @rawlingsj can give it a test though to be sure...

**Comment by rawlingsj** [comment]
The links from the toolbox in the top right work and so did the project dashbaord however there was one thing I noticed, the link from the dashbaord wasn't using the ingress external hostname but the external loadbalanced IP address instead.

I dont think the project dashboard is using the exposeURL annotations on the services.

**Comment by gashcrumb** [comment]
@rawlingsj ah lemme look, those must be calculating their own service link I guess.

**Comment by gashcrumb** [comment]
@rawlingsj does your service in your project have the exposeUrl annotation?  We'll fall back to calculating it if ti's not found.  Just checked here with a Camel integration project, the generate service didn't have this annotation, I added it manually and the links then from the team and project dashboards seem right for that service.

**Comment by rawlingsj** [comment]
Yeah it did, the toolbox link worked BTW.

Did you use a recent version of gofabric8?  The latest is 0.4.54 works with the exposecontroller:latest to update the annotations.

**Comment by gashcrumb** [comment]
Huh, so the jenkins and gogs URLs being used actually come from an annotation on the ReplicaSet object, at least on the team dashboard, @jstrachan are those normally populated by forge?  Lemme check the app dashboard.

**Comment by jstrachan** [comment]
Yeah the BuildConfig object gets annotated by the Jenkins + gogs URLs as part of the create project wizard. That logic comes from the DevOpsConnector + Java code. 

I thought I fixed that Java code to reuse the annotations too though...

**Comment by jstrachan** [comment]
Oh - the jenkins + gogs URLs get set when the project is created by Forge IIRC. Maybe re-create a project to see if the URLs get updated correctly?

**Comment by gashcrumb** [comment]
@jstrachan do I need an updated forge maybe?

**Comment by gashcrumb** [comment]
So for a local test I updated the annotation on the Jenkins service to be 'jenkins.fabric8' and aliased it in my /etc/hosts (well, the windows equiv).  I then created a new project, but it looks like the annotations in the generate ReplicaSet are still using the old exposeUrl value, could be maybe that gets cached?  I guess we could just let the console figure out the right build/git commit URLs instead and maybe simplify that code a bit, since the console has all the info it needs anyways.
